### ISSUE TITLE
Event infrastructure

### ISSUE THREAD

**Issue by Timoses** [Timoses/Caster#13]
Caster could provide an event infrastructure which allows not only Caster itself but plugins as well to emit events.

Original Caster uses a form of [Publish/Subscribe](https://en.wikipedia.org/wiki/Publish%E2%80%93subscribe_pattern) pattern which could also be useful for Caster3.

**Comment by Timoses** [comment]
We should also clearly work out what problems an event infrastructure would try to solve, e.g.:
* Informing the user of certain events
    * Still not clear how?
* Providing hooks/subscriptions to plugins
    * For what purpose?

What should also be considered is the possibility of Caster in a [client-server model](https://github.com/Timoses/Caster/issues/14). How would that impact event architecture?

**Comment by LexiconCode** [comment]
There is a lot that we could do here both with plug-ins itself and the guts of caster core. 

The basic idea is that Caster itself defines events, objects which contain **immutable** information about the stuff going on in the "guts" of Caster, and then those objects are fed to "hooks", user-defined listeners which can do whatever they want with said information. This should allow a decent amount of customization to happen without needing to make changes in the "guts".

Loosely coupled code leads to less bugs and more productivity, and so the use of hooks rather than "guts" modifications is encouraged whenever possible. Not only does is limit bugs very specific parts of the code but it allows the end-user to have minimal changes to any part of the CodeBase if they desire.

On see how they're used in the original caster framework. Here's an example of how they could work for Caster Core using the original framework as an example. 

Example Hooks with Events:

- [printer_hook.py](https://github.com/dictation-toolbox/Caster/blob/master/castervoice/lib/merge/ccrmerging2/hooks/examples/printer_hook.py) utilizes [activation event](https://github.com/dictation-toolbox/Caster/blob/master/castervoice/lib/merge/ccrmerging2/hooks/events/activation_event.py)
- [format_hook.py](https://github.com/dictation-toolbox/Caster/blob/master/castervoice/lib/merge/ccrmerging2/hooks/standard_hooks/format_hook.py) utilizes  [activation event](https://github.com/dictation-toolbox/Caster/blob/master/castervoice/lib/merge/ccrmerging2/hooks/events/activation_event.py)
- [node_change_hook.py](https://github.com/dictation-toolbox/Caster/blob/master/castervoice/lib/merge/ccrmerging2/hooks/events/node_change_event.py) utilizes [node change event](https://github.com/dictation-toolbox/Caster/blob/master/castervoice/lib/merge/ccrmerging2/hooks/events/node_change_event.py)
- [show_window_on_error_hook.py](https://github.com/dictation-toolbox/Caster/blob/5172a44d3cd58619f6228231e3aef2fddd1f1fb3/castervoice/lib/ctrl/mgr/grammar_manager.py#L302) utilizes [on error event](https://github.com/dictation-toolbox/Caster/blob/master/castervoice/lib/merge/ccrmerging2/hooks/events/on_error_event.py)

Hook Runner Placement placement in Caster source code:

-  `activation event` in [grammar_manager.py](https://github.com/dictation-toolbox/Caster/blob/3ff4f7d7c9c01fec2059ffa5c4ca708fdb7d09ad/castervoice/lib/ctrl/mgr/grammar_manager.py#L151)
-  `node change event` in [tree_rule.py](https://github.com/dictation-toolbox/Caster/blob/5172a44d3cd58619f6228231e3aef2fddd1f1fb3/castervoice/lib/merge/selfmod/tree_rule/tree_rule.py#L57)
-  `on error event` in [grammar_manager.py](https://github.com/dictation-toolbox/Caster/blob/5172a44d3cd58619f6228231e3aef2fddd1f1fb3/castervoice/lib/ctrl/mgr/grammar_manager.py#L302)

What we could do is before a grammar is loaded into the engine from a plug-in we want to do X. This is where we could begin to provide a framework to modify grammars from plug-ins via an API even leveraging existing code what we have for transformers without having to worry about merging.

However this can be expanded where plug-ins can register Events.

**Comment by Timoses** [comment]
Yes, plugins should be able to emit their own events which should be uniquely identified. Perhaps something like `casterplugin.dictation.events.OnDictationStart` is possible as a UID; being similar to how Plugins are currently identified.

Regarding client-server: I'm sure there's frameworks or suits out there which support such architecture designs involving events emitted over various devices / clients (and if there are non which supports our idea then it might be more likely that our ideas are off : ) ).

**Comment by Timoses** [comment]
The event infrastructure could also be used to create notifications for user-facing clients.
